pyomo/solvers/tests/checks/test_SAS.py

```diff
@@ -0,0 +1,455 @@
+        )
+
+
+class SASTestLPCAS(SASTestLP):
+    solver_io = '_sascas'
+    base_options = CAS_OPTIONS
+
+
+class SASTestMILP(SASTestAbc, unittest.TestCase):
+    def setX(self):
+        self.instance.X = Var([1, 2, 3], within=NonNegativeIntegers)
+
+    def checkSolution(self):
+        instance = self.instance
+        results = self.results
+
+        # Get the objective sense, we use the same code for minimization and maximization tests
+        sense = instance.Obj.sense
+
+        # Check status
+        self.assertEqual(results.solver.status, SolverStatus.ok)
+        self.assertEqual(
+            results.solver.termination_condition, TerminationCondition.optimal
+        )
+
+        # Check objective value
+        self.assertAlmostEqual(instance.Obj(), sense * -7)
+
+        # Check primal solution values
+        self.assertAlmostEqual(instance.X[1].value, 0.0)
+        self.assertAlmostEqual(instance.X[2].value, 1.0)
+        self.assertAlmostEqual(instance.X[3].value, 1.0)
+
+    @unittest.skipIf(not sas_available, "The SAS solver is not available")
+    def test_solver_default(self):
+        self.run_solver(options={})
+        self.checkSolution()
+
+    @unittest.skipIf(not sas_available, "The SAS solver is not available")
+    def test_solver_presolve(self):
+        self.run_solver(options={"presolver": "none"})
+        self.assertIn(
+            "NOTE: The MILP presolver value NONE is applied.", self.opt_sas._log
+        )
+        self.checkSolution()
+
+    @unittest.skipIf(not sas_available, "The SAS solver is not available")
+    def test_solver_intoption(self):
+        self.run_solver(options={"maxnodes": 20})
+        self.checkSolution()
+
+    @unittest.skipIf(not sas_available, "The SAS solver is not available")
+    def test_solver_invalidoption(self):
+        with self.assertRaisesRegex(ValueError, "syntax error"):
+            self.run_solver(options={"foo": "bar"})
+
+    @unittest.skipIf(not sas_available, "The SAS solver is not available")
+    def test_solver_max(self):
+        X = self.instance.X
+        self.instance.Obj.set_value(expr=-2 * X[1] + 3 * X[2] + 4 * X[3])
+        self.instance.Obj.sense = maximize
+        self.run_solver()
+        self.checkSolution()
+
+    @unittest.skipIf(not sas_available, "The SAS solver is not available")
+    def test_solver_infeasible(self):
+        instance = self.instance
+        X = instance.X
+        instance.R4 = Constraint(expr=-2 * X[2] - 3 * X[3] <= -6)
+        self.run_solver()
+        results = self.results
+        self.assertEqual(results.solver.status, SolverStatus.warning)
+        self.assertEqual(
+            results.solver.termination_condition, TerminationCondition.infeasible
+        )
+        self.assertEqual(results.solver.message, "The problem is infeasible.")
+
+    @unittest.skipIf(not sas_available, "The SAS solver is not available")
+    @unittest.skip("Returns wrong status for some versions.")
```

Comment:
If this test shows that there is an unexpected failure happening, can the failure be fixed?